Library/Homebrew/livecheck/strategy/cpan.rb

```diff
@@ -0,0 +1,67 @@
+# typed: false
+# frozen_string_literal: true
+
+require "uri"
+
+module Homebrew
+  module Livecheck
+    module Strategy
+      # The {Cpan} strategy identifies versions of software at
+      # cpan.metacpan.org by checking directory listing pages.
+      #
+      # CPAN URLs take the following format:
+      #
+      # * `https://cpan.metacpan.org/authors/id/M/MI/MIYAGAWA/Carton-v1.0.34.tar.gz`
+      #
+      # @api public
+      class Cpan
+        NICE_NAME = "CPAN"
+
+        # The allowlist used to determine if the strategy applies to the URL.
+        HOST_ALLOWLIST = ["cpan.metacpan.org"].freeze
+
+        # Whether the strategy can be applied to the provided URL.
+        #
+        # @param url [String] the URL to match against
+        # @return [Boolean]
+        def self.match?(url)
+          uri = URI.parse url
+          HOST_ALLOWLIST.include? uri.host
+        rescue URI::InvalidURIError
+          false
```

Comment:
I like this but I wonder if it would be better to keep the same approach as other strategies, that is
`URL_MATCH_REGEX.match?(url)` ? Currently (with a single host in host_allow_list) you'd have to do something like
```rb
URL_MATCH_REGEX = /cpan\.metacpan\.org/i.freeze
def self.match?(url)
  URL_MATCH_REGEX.match?(url)
end
```
I'm sure @samford can comment on that too.

Reply:
The `#match?` method here isn't sufficiently enforcing requirements, so we need to address that first. `#match?` should ensure that any matching URLs will be usable in `#find_versions`. That is to say, this is only ensuring that the host is `cpan.metacpan.org` and isn't making sure that the URL meets the requirements of the regex in `#find_versions` that identifies the author path and filename.

If we enforce those constraints, we end up with something more like:

```ruby
HOST_ALLOWLIST = ["cpan.metacpan.org"].freeze
URL_PATH_REGEX = %r{/authors/id(?:/[^/]+){3,}}i.freeze

def match?(url)
  begin
    uri = URI.parse url
  rescue URI::InvalidURIError
    return false
  end

  HOST_ALLOWLIST.include?(uri.host) && uri.path.match?(URL_PATH_REGEX)
end
```

For comparison, the `URL_MATCH_REGEX` approach would be something like:

```ruby
URL_MATCH_REGEX = %r{//cpan\.metacpan\.org/authors/id(?:/[^/]+){3,}}i.freeze

def self.match?(url)
  URL_MATCH_REGEX.match?(url)
end
```

All that `URI#parse` is buying us is separation of the parts of the URL (host, path, etc.). Considering that `URL_MATCH_REGEX` covers both host and path comparison and doesn't require us to load a library, I think it's heavily in favor of continuing to use our existing approach. By being careful with our regex, it doesn't end up being necessary to separate the parts of the URL and we can avoid using `URI#parse`.

For good measure, I profiled the `URI#parse` approach against the `URL_MATCH_REGEX` approach. Predictably, loading `URI` allocates quite a bit of memory (though we're already using it in `Livecheck#preprocess_url`, so kind of moot right now) and also makes `#match?` around 25x slower.

With all that in mind, I think the detriments of using `URI#parse` outweigh any benefits and we can achieve the same goals using the existing `URL_MATCH_REGEX` approach.

Reply:
I think it's better to maintain consistency and use the `URL_MATCH_REGEX` approach, unless this approach has added benefits I'm unaware of.

Reply:
> The `#match?` method here isn't sufficiently enforcing requirements, so we need to address that first.

That is very much true and it is probably a lousy move to try to justify it by following existing hackage, npm & pypi strategies. They are similarly insufficient and I decided to put consistency over correctness 🙃 
- `hackage.rb`
  ```rb
  # Hackage URLs take one of the following formats:
  #
  # * `https://hackage.haskell.org/package/example-1.2.3/example-1.2.3.tar.gz`
  # * `https://downloads.haskell.org/~ghc/8.10.1/ghc-8.10.1-src.tar.xz`
  #
  URL_MATCH_REGEX = /(?:downloads|hackage)\.haskell\.org/i.freeze

  def self.match?(url)
    URL_MATCH_REGEX.match?(url)
  end
  ```

- `npm.rb`
  ```rb
  # npm URLs take one of the following formats:
  #
  # * `https://registry.npmjs.org/example/-/example-1.2.3.tgz`
  # * `https://registry.npmjs.org/@example/example/-/example-1.2.3.tgz`
  #
  URL_MATCH_REGEX = /registry\.npmjs\.org/i.freeze

  def self.match?(url)
    URL_MATCH_REGEX.match?(url)
  end
  ```

- `pypi.rb`
  ```rb
  # PyPI URLs have a standard format but the hexadecimal text between
  # `/packages/` and the filename varies:
  #
  # * `https://files.pythonhosted.org/packages/<hex>/<hex>/<long_hex>/example-1.2.3.tar.gz`
  #
  URL_MATCH_REGEX = /files\.pythonhosted\.org/i.freeze

  def self.match?(url)
    URL_MATCH_REGEX.match?(url)
  end
  ```

> For good measure, I profiled the `URI#parse` approach against the `URL_MATCH_REGEX` approach. `URI#parse` involves more memory allocations and also makes `Cpan#match?` around 25x slower. `#match?` is run for every strategy across every formula that livecheck is checking, so this performance issue would add up.
> 
> With all that in mind, I think the detriments of using `URI#parse` outweigh any benefits and we can achieve the same goals using the existing `URL_MATCH_REGEX` approach.

While I agree with the expressed sentiment and can't argue with facts (performance analysis) note that you need to _anchor_ that regex to fully replicate safety guarantees provided by URI parser:
```rb
URL_MATCH_REGEX = %r{^https?://cpan\.metacpan\.org/authors/id(?:/[^/]+){3,}/[^/]+}i.freeze
```
otherwise, it is possible (although highly unlikely) to get false positives:
- https://rubular.com/r/dpi5tfTUiRtlNm (correct)
- https://rubular.com/r/xEBvSiHyMnGJPY (incorrect)

Reply:
> `Hackage`, `Npm`, `Pypi`...

Yep, the `#match?` methods in those strategies could do with being updated in the same fashion. I'll make a note so I remember to address those in a follow-up PR.

> `URL_MATCH_REGEX = %r{^https?://cpan\.metacpan\.org/authors/id(?:/[^/]+){3,}/[^/]+}i.freeze`

I was thinking about being explicit and starting the regex with `^https?://` instead of `//`, so I'm good with this 👍

Reply:
> to address those in a follow-up PR.

I'll do it before it vanishes from my cache 😄